### PR TITLE
Lottie Slot Feature Support

### PR DESCRIPTION
**Pull request by tinyjin**
Internal model & parser changes to parse "sid" and obtain their data into the `LottieComposition`.
It will support dynamic-change following lottie objects:
- LottieSolidStroke
- LottieSolidFill
- LottieGradientStroke
- LottieGradientFill
- LottieTextDoc

Related: #1808

### CODE REVIEW

File: src/loaders/lottie/tvgLottieModel.h [hunk 1]
```diff
@@ -456,7 +457,10 @@ struct LottieSolidFill : LottieObject
 };
 
 
-struct LottieGradientFill : LottieObject, LottieGradient
+struct LottieGradientObject : LottieObject, LottieGradient {};
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieModel.h:460]: Can't see any necessity of LottieGradientObject in this commit context. Could you please explain this?

Comment (tinyjin) [reply src/loaders/lottie/tvgLottieModel.h:460]: In this change, `parseProperty` will have parameter `LottieObject` to parse `sid` into itself.

When system calls `parseProperty` from `parseGradient`:
```cpp
void LottieParser::parseGradient(LottieGradient* gradient, const char* key)
{
    // ...
    // Can't pass original lottie object because LottieGradient isn't child of LottieObject for this scope.
    else if (!strcmp(key, "o")) parseProperty(gradient->opacity, gradient); // [!] Type error
    // ...
}
```

Even though, `parseGradient` is called with LottieObject such as
- LottieGradientFill
- LottieGradientStroke

the parameter is being type-changed because of the explicit type of `LottieGradient`, it wasn't able to pass as `LottieObject`.


But if we declare `LottieGradientObject`, the `LottieObject` will be correctly passed.
```cpp
// the `gradient` is matched the `LottieObject`
void LottieParser::parseGradient(LottieGradientObject* gradient, const char* key)

// template<typename T> void parseProperty(T& prop, LottieObject* parent)
```

**Review by hermet — COMMENTED** [review]
Please check my comments.

File: src/loaders/lottie/tvgLottieModel.h [hunk 2]
```diff
@@ -618,6 +639,7 @@ struct LottieComposition
     Array<LottieObject*> assets;
     Array<LottieInterpolator*> interpolators;
     Array<LottieFont*> fonts;
+    Array<LottieSlot*> slots;
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieModel.h:642]: the slots in the array should be freed at ~LottieComposition();

**Review by hermet — COMMENTED** [review]
Thanks, one more comment; everything else looks fine so far. I'm looking for the following implementation patches.

File: src/loaders/lottie/tvgLottieParser.cpp [hunk 3]
```diff
@@ -468,12 +468,20 @@ void LottieParser::parsePropertyInternal(T& prop)
 
 
 template<typename T>
-void LottieParser::parseProperty(T& prop)
+void LottieParser::parseProperty(T& prop, LottieSlot::Type type, LottieObject* parent)
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieParser.cpp:472]: Considering again, the term 'parent' here, does it function conceptually as a parent in a hierarchical structure, or is it simply the 'object' associated with a property? I thin it's more likely to be => LottieObject* obj

**Review by hermet — COMMENTED** [review]
Reviewed: "lottie/parser: Added sid parsing logic"

File: inc/thorvg.h [hunk 4]
```diff
@@ -1769,6 +1769,18 @@ class TVG_API Animation
...
+     * @brief Rewrite lottie properties through the slot data
+     *
+     * @param[in] slotJson The lottie slot data
+     *
+     * @retval Result::Success When succeed.
+     * @retval Result::InsufficientCondition When the given parameter is invalid.
+     *
+     * @note Experimental API
+     */
+    Result override(const char* slotJson) noexcept;
```
Comment (hermet) [inline inc/thorvg.h:1782]: This concept is not suitable for general Animation. Please remove the API implementation from this patch. We can consider introducing LottieAnimation instead.

File: src/loaders/lottie/tvgLottieOverrider.h [hunk 5]
```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright (c) 2023 - 2024 the ThorVG project. All rights reserved.
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieOverrider.h:2]: -> 2024

File: src/loaders/lottie/tvgLottieOverrider.cpp [hunk 6]
```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (c) 2023 - 2024 the ThorVG project. All rights reserved.
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieOverrider.cpp:2]: -> 2024

File: src/loaders/lottie/tvgLottieModel.h [hunk 7]
```diff
@@ -613,6 +613,28 @@ struct LottieComposition
 {
     ~LottieComposition();
 
+    LottieObject* findObject(const char* sid) {
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieModel.h:616]: Suggestion: 
Essentially, ThorVG prefers simpler method names (if no any conflict) that consist of just a verb, like.
```
LottieObject* find(const char* sid);
```
As you can see, we can easily figure out the purpose of this method through method name and the return type value + the given parameter.

File: src/loaders/lottie/tvgLottieModel.h [hunk 8]
```diff
@@ -613,6 +613,28 @@ struct LottieComposition
...
+        auto targets = this->root->children;
+        return getObject(sid, root->children);
+    }
+
+    LottieObject* getObject(const char* sid, const Array<LottieObject*> children) {
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieModel.h:621]: Should this method be introduced as a member of LottieComposition()? I think this is not relevant to it. Probably you can add this as a just static helper function (see 29 line in tvgLottieModel.cpp )

Also, Just for your information, Please double-check the difference in binary size between inline (.h) and (.cpp) implementations when you adding a inline function. We will use the one with the smaller size.

File: src/renderer/tvgAnimation.cpp [hunk 9]
```diff
@@ -24,6 +24,8 @@
 #include "tvgFrameModule.h"
```
Comment (hermet) [inline src/renderer/tvgAnimation.cpp:24]: Let's make a Animation change with a separate commit.

File: src/loaders/lottie/tvgLottieParser.h [hunk 10]
```diff
@@ -35,6 +35,7 @@ struct LottieParser : LookaheadParserHandler
     }
 
     bool parse();
+    bool parseSlot(LottieComposition* comp, LottieSlot* slot);
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieParser.h:38]: We can remove the 'comp' parameter since this parser already has a designated 'comp' (see line 40).
``` parseSlot() -> parse(LottieSlot* slot);  ```
However, I expect the ```parse(const char* slot)```,  parsing json logic sholud be in the parser. We can eliminate the dependency on RapidJSON from the overrider if the parsing tasks can be transferred to LottieParser.

File: src/loaders/lottie/tvgLottieParser.cpp [hunk 11]
```diff
@@ -1243,6 +1244,67 @@ void LottieParser::postProcess(Array<LottieGlyph*>& glyphes)
 /* External Class Implementation                                        */
 /************************************************************************/
 
+bool LottieParser::parseSlot(LottieComposition* comp, LottieSlot* slot)
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieParser.cpp:1247]: please double check indentaion.

File: src/loaders/lottie/tvgLottieOverrider.cpp [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+#include "tvgLottieParser.h"
+#include "tvgLottieLoader.h"
+
+using namespace std;
+using namespace rapidjson;
+
+/************************************************************************/
+/* Internal Class Implementation                                        */
+/************************************************************************/
+
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieOverrider.cpp:39]: please double check indentation.

File: src/renderer/tvgAnimation.cpp [hunk 13]
```diff
@@ -62,6 +65,21 @@ Animation::Animation() : pImpl(new Impl)
 }
 
 
+Result Animation::override(const char* slotJson) noexcept
+{
+    if (!pImpl->overrider) {
+        auto loader = static_cast<LottieLoader*>(pImpl->picture->pImpl->loader);
```
Comment (hermet) [inline src/renderer/tvgAnimation.cpp:71]: This should be something like this:
```
static_cast<LottieLoader*>(pImpl->picture->pImpl->loader)->override(slot);
```
Then, the LottieOverrider can reside within LottieLoader.

File: src/loaders/lottie/tvgLottieParser.cpp [hunk 14]
```diff
@@ -1243,6 +1244,67 @@ void LottieParser::postProcess(Array<LottieGlyph*>& glyphes)
...
+    enterObject();
+
+    // Generate context
+    LottieParser::Context context;
+    this->context = &context;
+
+    // Find target parent object
+    auto target = comp->findObject(slot->sid);
+
+    switch (slot->type)
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieParser.cpp:1261]: please check a coding convention about switch() at 374 line tvgSwShape.cpp

**Review by hermet — CHANGES_REQUESTED** [review]
reviewed "lottie: Support the slot overriding feature"

File: src/examples/LottieSlot.cpp [hunk 15]
```diff
@@ -0,0 +1,227 @@
+/*
+ * Copyright (c) 2023 - 2024 the ThorVG project. All rights reserved.
```
Comment (hermet) [inline src/examples/LottieSlot.cpp:2]: 2024

**Review by hermet — COMMENTED** [review]
reviewd "examples: Added lottie slot example"

File: src/loaders/lottie/tvgLottieParser.cpp (hunk 3 above)
Comment (tinyjin) [reply src/loaders/lottie/tvgLottieParser.cpp]: Yes, it's just associated object, renamed it

File: src/loaders/lottie/tvgLottieModel.cpp [hunk 16]
```diff
@@ -233,4 +249,9 @@ LottieComposition::~LottieComposition()
     for (auto s = slots.data; s < slots.end(); ++s) {
         delete(*s);
     }
+}
+
+LottieObject* LottieComposition::find(const char* sid) {
+    auto targets = this->root->children;
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieModel.cpp:255]: one liner:
```
    return _getObject(sid, this->root->children);
```

File: src/loaders/lottie/tvgLottieModel.cpp [hunk 17]
```diff
@@ -29,6 +29,22 @@
 /* Internal Class Implementation                                        */
 /************************************************************************/
 
+static LottieObject* _getObject(const char* sid, const Array<LottieObject*> children) {
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieModel.cpp:32]: Avoid memory copying:
```
const Array<LottieObject*> children -> , const Array<LottieObject*>& children
```

File: src/loaders/lottie/tvgLottieOverrider.cpp [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+    parser->comp = this->comp;
+
+    auto sid = parser->getSid();
+    if (!sid) return false;
+
+    auto slots = comp->slots;
+    for (auto s = slots.data; s < slots.end(); ++s) {
+        auto slot = (*s);
+
+        if (!strcmp(slot->sid, sid)) {
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieOverrider.cpp:51]:
```
if (!strcmp(slot->sid, sid)) return parser->parse(slot);
```

File: src/loaders/lottie/tvgLottieOverrider.cpp [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+            parser->parse(slot);
+            break;
+        }
+    }
+
+    return true;
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieOverrider.cpp:57]: probably false?

File: src/loaders/lottie/tvgLottieOverrider.cpp [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+
+void LottieOverrider::overrideGradient(LottieObject* target, LottieGradient* gradient)
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieOverrider.cpp:61]: Let's use a simple name (override) since we can easily figure out the objective with the parameters.

File: src/loaders/lottie/tvgLottieOverrider.cpp [hunk 21]
```diff
@@ -0,0 +1,115 @@
...
+{
+    switch (target->type)
+    {
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieOverrider.cpp:64]: Please double check the convention:
image

File: src/loaders/lottie/tvgLottieOverrider.cpp [hunk 22]
```diff
@@ -0,0 +1,115 @@
...
+        TVGERR("LOTTIE", "Invalid target type for gradient override!");
+        break;
+    }
+}
+
+
+void LottieOverrider::overrideColor(LottieObject* target, LottieObject* color)
+{
+    auto solid = static_cast<LottieSolidFill*>(color);
+    switch (target->type)
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieOverrider.cpp:88]: fix convention:

File: src/loaders/lottie/tvgLottieOverrider.cpp [hunk 23]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+
+void LottieOverrider::overrideGradient(LottieObject* target, LottieGradient* gradient)
+{
+    switch (target->type)
+    {
+        case LottieObject::GradientFill: {
+            auto fill = static_cast<LottieGradientFill*>(target);
+            fill->colorStops = gradient->colorStops;
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieOverrider.cpp:67]: 1. Maybe it needs to free the ex-memory... ```fill->colorStops->frames``` overriding  operator=() ?
2. It seems here polymorphism helps to make code neat & clean. Please consider the interface of the next:
```
virtual bool LottieObject::override(LottieObject* prop) {}

bool LottieGradientFill::override(LottieObject* prop) override
{
    auto fill = static_cast<LottieGradientFill*>(prop);
    this->colorStops = gradient->ColorStops;
}

```

**Review by hermet — CHANGES_REQUESTED** [review]
Reviewed: "lottie: Support the slot overriding feature"

File: src/loaders/lottie/tvgLottieAnimation.cpp [hunk 24]
```diff
@@ -0,0 +1,55 @@
...
+/************************************************************************/
+
+LottieAnimation::~LottieAnimation()
+{
+    delete(pImpl);
+}
+
+Result LottieAnimation::override(const char* slotJson) noexcept
+{
+    if (static_cast<LottieLoader*>(pImpl->picture->pImpl->loader)->override(slotJson)) {
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieAnimation.cpp:44]: if (!pImpl->picture->pImpl->loader) return InsufficientCondition;

File: src/renderer/tvgAnimation.cpp [hunk 25]
```diff
@@ -20,33 +20,15 @@
  * SOFTWARE.
  */
 
-#include "tvgCommon.h"
 #include "tvgFrameModule.h"
-#include "tvgPaint.h"
 #include "tvgPicture.h"
+#include "tvgLottieLoader.h"
```
Comment (hermet) [inline src/renderer/tvgAnimation.cpp:25]: this file doesn't refer: 
-- #include "tvgLottieLoader.h"

-- #include "tvgPicture.h"   //can be included in tvgAnimation.h

File: src/renderer/tvgAnimation.h [hunk 26]
```diff
@@ -0,0 +1,47 @@
...
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifndef _TVG_ANIMATION_H_
+#define _TVG_ANIMATION_H_
+
+#include "tvgCommon.h"
+#include "tvgPaint.h"
```
Comment (hermet) [inline src/renderer/tvgAnimation.h:27]: ++ #include "tvgPicture.h"

File: inc/thorvg.h [hunk 27]
```diff
@@ -1849,6 +1849,45 @@ class TVG_API Animation
...
```
Comment (hermet) [inline inc/thorvg.h:1878]: slotJson -> slot

File: inc/thorvg.h [hunk 28]
```diff
@@ -1849,6 +1849,45 @@ class TVG_API Animation
...
+    ~LottieAnimation();
+
+    /**
+     * @brief Rewrite lottie properties through the slot data
+     *
+     * @param[in] slotJson The lottie slot data
+     *
+     * @retval Result::Success When succeed.
+     * @retval Result::InsufficientCondition When the given parameter is invalid.
+     *
```
Comment (hermet) [inline inc/thorvg.h:1875]: ++InsufficientCondition

**Review by hermet — CHANGES_REQUESTED** [review]
reviewd "lottie: Introduce the LottieAnimation class"

File: src/examples/LottieSlot.cpp [hunk 29]
```diff
@@ -0,0 +1,227 @@
...
+    auto shape = tvg::Shape::gen();
+    shape->appendRect(0, 0, WIDTH, HEIGHT);
+    shape->fill(50, 50, 50);
+
+    if (canvas->push(std::move(shape)) != tvg::Result::Success) return;
+
+    const char* slotData = R"(
+    {
+        "gradient_fill": {
+            "p": {
```
Comment (hermet) [inline src/examples/LottieSlot.cpp:57]: this json is ok?

**Comment by hermet** [comment]
@tinyjin thanks for your update, please check my comments.

**Comment by hermet** [comment]
Imagine one another scenario, if the user intends to cancel the override of the slot by using:
```
animation->override(nullptr);
```

File: src/renderer/tvgAnimation.cpp (hunk 25 above)
Comment (tinyjin) [reply src/renderer/tvgAnimation.cpp:25]: Without `tvgLottieLoader.h`, the `LottieLoader` cannot be found

> identifier "LottieLoader" is undefinedC/C++(20)
```cpp
if (static_cast<LottieLoader*>(pImpl->picture->pImpl->loader)->override(slot)) {
```

File: inc/thorvg.h (hunk 28 above)
Comment (tinyjin) [reply inc/thorvg.h:1875]: line: 1874
`InsufficientCondition` is already there?

File: src/renderer/tvgAnimation.cpp (hunk 25 above)
Comment (hermet) [reply src/renderer/tvgAnimation.cpp:25]: Please double-check it, I couldn't see any LottieLoader usage in tvgAnimation.cpp.
Even the call is belonged to tvgLottieAnimation.cpp, not tvgAnimation.cpp

File: src/loaders/lottie/tvgLottieOverrider.cpp [hunk 30]
```diff
@@ -0,0 +1,115 @@
...
+ * SOFTWARE.
+ */
+
+#include "tvgLottieOverrider.h"
+#include "tvgCommon.h"
+#include "tvgLottieOverrider.h"
+#include "tvgLottieParser.h"
+#include "tvgLottieLoader.h"
+
+using namespace std;
```
[the remark this answers is not in the document]
Comment (hermet) [reply src/loaders/lottie/tvgLottieOverrider.cpp:29]: I meant the 29 line.

File: src/renderer/tvgAnimation.cpp (hunk 25 above)
Comment (tinyjin) [reply src/renderer/tvgAnimation.cpp:25]: Oops, I checked wrong file. that's correct.

File: src/examples/LottieSlot.cpp (hunk 29 above)
Comment (tinyjin) [reply src/examples/LottieSlot.cpp:57]: Screenshot 2024-02-21 at 1 00 42 AM
Yes, In some white-space cases, RapidJSON cannot parse, but this example JSON can be parsed.

We might need to find the crashed pattern with the white-space.

**Comment by hermet** [comment]
> > Imagine one another scenario, if the user intends to cancel the override of the slot by using:
> > ```
> > animation->override(nullptr);
> > ```
> 
> @hermet Sounds good.
> 
> And wondering:
> 
> * We discussed this overriding features shouldn't take a memory to record what's overrided.
> * Do we have efficient way to clear overridden status without saving changes?

Please check this: https://github.com/thorvg/thorvg/pull/1998

```
LottieSlot
{
  LottieProperty::Type type;
  char* sid;
  LottieProperty* property  //we can backup the origin here:
};
```
Probably you can make the slot rewind the sequence with the original property in LottiSlot.

File: src/loaders/lottie/tvgLottieModel.h [hunk 31]
```diff
@@ -590,6 +592,25 @@ struct LottieLayer : LottieGroup
 };
 
 
+struct LottieSlot
+{
+    char* sid = nullptr;
+
+    enum Type : uint8_t
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieModel.h:631]: Could you please cherry pick this first, and make this based on it?
https://github.com/thorvg/thorvg/pull/1998

File: src/loaders/lottie/tvgLottieModel.cpp [hunk 32]
```diff
@@ -228,4 +228,9 @@ LottieComposition::~LottieComposition()
     for (auto f = fonts.begin(); f < fonts.end(); ++f) {
         delete(*f);
     }
+
+    //delete slots
+    for (auto s = slots.begin(); s < slots.end(); ++s) {
+        delete(*s);
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieModel.cpp:248]: Considering again, freeing the resource by Slot is better here instead of 124 line.
```
free(s->sid)
```

File: src/loaders/lottie/tvgLottieModel.h [hunk 33]
```diff
@@ -620,6 +641,7 @@ struct LottieComposition
     Array<LottieObject*> assets;
     Array<LottieInterpolator*> interpolators;
     Array<LottieFont*> fonts;
+    Array<LottieSlot*> slots;
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieModel.h:678]: Reviewing again, there is no point to use dynamic memory allocation for slots. 
```
Array<LottieSlot*> slots -> Array<LottieSlot> slots
```

File: src/loaders/lottie/tvgLottieParser.cpp [hunk 34]
```diff
@@ -457,12 +457,20 @@ void LottieParser::parsePropertyInternal(T& prop)
 
 
 template<typename T>
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieParser.cpp:460]: Probably, we can remove a parameter, by adding: 
```
template<typename T, LottieProperty type>
void LottieParser::parseProperty(T& prop, LottieObject* obj)
```

See https://github.com/thorvg/thorvg/pull/1998

File: src/loaders/lottie/tvgLottieParser.cpp [hunk 35]
```diff
@@ -1228,6 +1229,77 @@ void LottieParser::postProcess(Array<LottieGlyph*>& glyphes)
...
+            auto gradient = new LottieGradient;
+            this->context->gradient = gradient;
+
+            while (auto key = nextObjectKey()) {
+                if (!strcmp(key, "p")) parseProperty(gradient->colorStops);
+                else skip(key);
+            }
+
+            LottieOverrider::override(targetObj, gradient);
+            delete(gradient);
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieParser.cpp:1264]: assuming this memory deletion also delete the property that has been belonged to the targetObj. So memory corruption... Please double check,

File: src/loaders/lottie/tvgLottieOverrider.h [hunk 36]
```diff
@@ -0,0 +1,43 @@
...
+#define _TVG_LOTTIE_OVERRIDER_H_
+
+#include "tvgCommon.h"
+#include "tvgLottieModel.h"
+
+struct LottieOverrider
+{
+    bool override(const char* data);
+
+    static void override(LottieObject* target, LottieGradient* gradient);
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieOverrider.h:33]: these 3 override() looks unnecessary...

File: src/loaders/lottie/tvgLottieParser.cpp [hunk 37]
```diff
@@ -1228,6 +1229,77 @@ void LottieParser::postProcess(Array<LottieGlyph*>& glyphes)
...
+                if (!strcmp(key, "p")) parseProperty(text->doc);
+                else skip(key);
+            }
+
+            LottieOverrider::override(targetObj, text);
+            delete(text);
+            break;
+        }
+        default: {
+            TVGERR("LOTTIE", "Unsupported slot type");
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieParser.cpp:1292]: You can put this log message in LottieObject::override()...

File: src/loaders/lottie/tvgLottieModel.h [hunk 38]
```diff
@@ -468,6 +488,12 @@ struct LottieGradientFill : LottieGradient
         if (LottieGradient::prepare()) statical = false;
     }
 
+    void override(LottieObject* prop) override
+    {
+        this->colorStops = static_cast<LottieGradient*>(prop)->colorStops;
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieModel.h:493]: definitely memory leak due to the previous property value.

File: src/loaders/lottie/tvgLottieModel.h [hunk 39]
```diff
@@ -350,6 +364,12 @@ struct LottieSolidFill : LottieSolid
         if (color.frames || opacity.frames) statical = false;
     }
 
+    void override(LottieObject* prop) override
+    {
+        this->color = static_cast<LottieSolid*>(prop)->color;
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieModel.h:369]: definitely memory leak due to the previous property value.

File: src/loaders/lottie/tvgLottieModel.h [hunk 40]
```diff
@@ -339,6 +347,12 @@ struct LottieSolidStroke : LottieSolid, LottieStroke
         LottieObject::type = LottieObject::SolidStroke;
         if (color.frames || opacity.frames || LottieStroke::dynamic()) statical = false;
     }
+
+    void override(LottieObject* prop) override
+    {
+        this->color = static_cast<LottieSolid*>(prop)->color;
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieModel.h:353]: definitely memory leak due to the previous property value.

File: src/loaders/lottie/tvgLottieModel.h [hunk 41]
```diff
@@ -183,6 +185,12 @@ struct LottieText : LottieObject
         LottieObject::type = LottieObject::Text;
     }
 
+    void override(LottieObject* prop) override
+    {
+        this->doc = static_cast<LottieText*>(prop)->doc;
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieModel.h:190]: definitely memory leak due to the previous property value.

**Review by hermet — CHANGES_REQUESTED** [review]
reviewed "lottie: Support the slot overriding feature"

File: src/loaders/lottie/tvgLottieModel.h [hunk 42]
```diff
@@ -121,9 +121,11 @@ struct LottieObject
     virtual ~LottieObject()
     {
         free(name);
+        free(sid);
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieModel.h:124]: doesn't this sid double free? => 234 line

File: src/loaders/lottie/tvgLottieParser.cpp [hunk 43]
```diff
@@ -456,13 +456,21 @@ void LottieParser::parsePropertyInternal(T& prop)
...
         if (!strcmp(key, "k")) parsePropertyInternal(prop);
-        else skip(key);
+        else if (obj && !strcmp(key, "sid")) {
+            auto slot = new LottieSlot;
+            if (!slot) return;
+
+            slot->sid = getStringCopy();
+            obj->sid = slot->sid;
+            slot->type = type;
+            comp->slots.push(*slot);
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieParser.cpp:473]: Please check the next, comp.slots doesn't have elements with the memory pointer anymore.
```
comp->slots.push({getStringCopy(), type});
obj->sid = comp->slots.last().sid;
```

**Review by hermet — CHANGES_REQUESTED** [review]
reviewed "lottie/parser: Added sid parsing logic"

File: src/loaders/lottie/tvgLottieModel.cpp [hunk 44]
```diff
@@ -29,6 +29,20 @@
...
+static LottieObject* _getObject(const char* sid, const Array<LottieObject*>& children) {
+    for (auto c = children.begin(); c < children.end(); ++c) {
+        auto child = *c;
+        if (child->type == LottieObject::Type::Group || child->type == LottieObject::Type::Layer) {
+            auto group = static_cast<LottieGroup*>(child);
+            auto foundObj = _getObject(sid, group->children);
+            if (foundObj) return foundObj;
+        }
+
+        if (child->sid && !strcmp(sid, child->sid)) return child;
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieModel.cpp:41]: This condition should be at 34 ~ 35 line, if the searching is at group itself?

File: src/loaders/lottie/tvgLottieModel.h [hunk 45]
```diff
@@ -124,6 +124,10 @@ struct LottieObject
         free(sid);
     }
 
+    virtual void override(LottieObject* prop) {
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieModel.h:127]: convention fix:
```
virtual void override(LottieObject* prop)
{
 ...
}
```

File: src/loaders/lottie/tvgLottieOverrider.cpp [hunk 46]
```diff
@@ -0,0 +1,50 @@
...
+
+/************************************************************************/
+/* Internal Class Implementation                                        */
+/************************************************************************/
+
+/************************************************************************/
+/* External Class Implementation                                        */
+/************************************************************************/
+
+bool LottieOverrider::override(const char* slot)
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieOverrider.cpp:35]: LottieOverride doesn't need to hold dirName / comp as member variables?
```
bool LottieOverrider::override(LottieComposition* comp, const char* slot, const char* dirName);
```

File: src/loaders/lottie/tvgLottieProperty.h [hunk 47]
```diff
@@ -249,6 +249,19 @@ struct LottieGenericProperty : LottieProperty
         return frame->interpolate(frame + 1, frameNo);
     }
 
+    T& operator=(const T& other)
+    {
+        if (this != &other) {
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieProperty.h:254]: Let's remove 254 condition. (we don't expect this scenario)

File: src/loaders/lottie/tvgLottieProperty.h [hunk 48]
```diff
@@ -249,6 +249,19 @@ struct LottieGenericProperty : LottieProperty
...
+            value = other.value;
+            if (other.frames) {
+                memcpy(frames, other.frames, sizeof(LottieScalarFrame<T>) * other.frames->count);
+                delete(other.frames);
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieProperty.h:258]: i think it needs to delete frames of "this" before memory copy, not "others".

**Review by hermet — CHANGES_REQUESTED** [review]
reviewed "lottie: Support the slot overriding feature"

**Comment by hermet** [comment]
@tinyjin Please check more comments. thanks!

File: src/loaders/lottie/tvgLottieModel.h [hunk 49]
```diff
@@ -124,6 +124,7 @@ struct LottieObject
     }
 
     char* name = nullptr;
+    char* sid = nullptr;
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieModel.h:132]: Could you please double check this?
```
Slot {
   char* sid;
   type;
   Array<LottieObject*> objs;   //assuming multiple properties uses the same sid.
};
```
I think LottieObject doesn't need to have sid member.

File: src/loaders/lottie/tvgLottieParser.cpp [hunk 50]
```diff
@@ -1223,6 +1224,76 @@ void LottieParser::postProcess(Array<LottieGlyph*>& glyphes)
...
+bool LottieParser::parse(LottieSlot* slot)
+{
+    enterObject();
+
+    // Generate context
+    LottieParser::Context context;
+    this->context = &context;
+
+    // Find target object
+    auto targetObj = comp->find(slot->sid);
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieParser.cpp:1246]: you can quickly find the targetObjs from the slots objs.

File: src/loaders/lottie/tvgLottieParser.cpp [hunk 51]
```diff
@@ -1223,6 +1224,76 @@ void LottieParser::postProcess(Array<LottieGlyph*>& glyphes)
...
+        case LottieProperty::Type::ColorStop: {
+            auto gradient = new LottieGradient;
+            this->context->gradient = gradient;
+
+            while (auto key = nextObjectKey()) {
+                if (!strcmp(key, "p")) parseProperty(gradient->colorStops);
+                else skip(key);
+            }
+
+            targetObj->override(gradient);
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieParser.cpp:1258]: And like this:
```
for (obj = slot.objs.begin(); obj < slot.objs.end(); ++obj) {
   (*obj)->override(gradient);
}
```

**Review by hermet — COMMENTED** [review]
Almost there, please check once more for the optimal way.

**Comment by hermet** [comment]
+please rebase the branch as well.

File: src/loaders/lottie/tvgLottieOverrider.h [hunk 52]
```diff
@@ -0,0 +1,34 @@
...
+
+#ifndef _TVG_LOTTIE_OVERRIDER_H_
+#define _TVG_LOTTIE_OVERRIDER_H_
+
+#include "tvgCommon.h"
+#include "tvgLottieModel.h"
+
+struct LottieOverrider
+{
+    bool override(LottieComposition* comp, const char* slot, const char* dirName);
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieOverrider.h:31]: Now, LottieOverrider is too simple, 
Let's migrate the LottieOverrider function with 
```LottieLoader::override(LottieComposition* comp, const char* slot, const char* dirName);```
and then remove LottieOverrider.

**Comment by hermet** [comment]
https://github.com/thorvg/thorvg/pull/2003
